### PR TITLE
add ability to not send templates based on related objects

### PR DESCRIPTION
**Pull request by Formulka**

### CODE REVIEW

File: pymess/migrations/0006_auto_20190322_1712.py [hunk 1]
```diff
@@ -1,40 +0,0 @@
...
-            model_name='dialermessage',
-            name='is_final_state',
-            field=models.BooleanField(default=False, verbose_name='is final state'),
-        ),
-        migrations.AlterField(
-            model_name='pushnotificationtemplate',
-            name='is_active',
-            field=models.BooleanField(default=True, verbose_name='is active'),
-        ),
-    ]
```
Comment (matllubos) [inline pymess/migrations/0006_auto_20190322_1712.py]: Mazat migraci je nebezpečné

File: pymess/models/common.py [hunk 2]
```diff
@@ -95,14 +110,22 @@ def clean_body(self, context_data=None):
...
 
-    def can_send(self, recipient, context_data):
-        return self.is_active
+    def can_send_for_object(self, related_objects):
+        if hasattr(self, 'disallowed_objects'):
+            return (
+                not self.disallowed_objects.filter_from_related_objects(related_objects).exists()
+            ) if related_objects is not None else True
+        else:
+            return True
```
Comment (matllubos) [inline pymess/models/common.py]: if else pattern?

`
return not hasattr(self, 'disallowed_objects') or related_objects is None or  not self.disallowed_objects.filter_from_related_objects(related_objects).exists()
`

File: pymess/models/sms.py [hunk 3]
```diff
@@ -75,3 +76,19 @@ class Meta(BaseAbstractTemplate.Meta):
 
 class SMSTemplate(AbstractSMSTemplate):
     pass
+
+
+class SMSTemplateDisallowedObject(BaseRelatedObject):
```
Comment (matllubos) [inline pymess/models/sms.py]: u dialeru to nemáme?

File: pymess/utils/html.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from typing import Optional, Sequence
+
+from bs4 import BeautifulSoup
+
+from django.core.exceptions import ValidationError
+from django.utils.translation import ugettext as _
+
+
+BANNED_TAGS = ('object', 'script', 'canvas', 'video', 'applet', 'embed', 'noscript', 'amp-iframe', 'application')
```
Comment (matllubos) [inline pymess/utils/html.py]: dej do settings prosím

File: pymess/models/sms.py (hunk 3 above)
Comment (Formulka) [reply pymess/models/sms.py]: však to tam je

File: pymess/utils/html.py (hunk 4 above)
Comment (Formulka) [reply pymess/utils/html.py]: jo to jsem chtěl udělat a zapomněl, díky

File: pymess/models/common.py (hunk 2 above)
Comment (Formulka) [reply pymess/models/common.py]: bývalo to v průběhu vývoje o dost složitější, díky

File: pymess/models/common.py [hunk 5]
```diff
@@ -29,6 +34,48 @@ def create_from_related_objects(self, *related_objects):
             for related_object in related_objects
         ]
 
+    def _get_related_object_pks(self, *related_objects):
+        return [related_object.pk for related_object in related_objects]
```
Comment (matllubos) [inline pymess/models/common.py]: where is it used?

File: pymess/models/common.py [hunk 6]
```diff
@@ -29,6 +34,48 @@ def create_from_related_objects(self, *related_objects):
...
+
+    def _get_related_objects_qs_kwargs(self, *related_objects):
+        return reduce(OR, (
+            Q(**{
+                'object_id': obj.pk,
+                'content_type': ContentType.objects.get_for_model(obj)
+            }) for obj in related_objects
+        ))
+
+    def update_related_objects(self, *related_objects, limit_to_model=None):
```
Comment (matllubos) [inline pymess/models/common.py]: when limit_to_model is None and you send

update_related_objects(related_object)

all other relations will be removed? It seems to be insecure

File: pymess/models/common.py [hunk 7]
```diff
@@ -29,6 +34,48 @@ def create_from_related_objects(self, *related_objects):
...
+        )
+
+    def get_or_create_from_related_objects(self, *related_objects):
+        return [
+            self.get_or_create_from_related_object(related_object)
+            for related_object in related_objects
+        ]
+
+    def get_related_objects_by_model(self, model):
+        pks = self.filter(content_type=ContentType.objects.get_for_model(model)).values_list('object_id_int', flat=True)
```
Comment (matllubos) [inline pymess/models/common.py]: values_list('object_id', flat=True)

object_id_int is set only for integer like primary keys

Comment (matllubos) [reply pymess/models/common.py]: is not necessary to use flat. Next queryset will work without it

File: pymess/models/common.py [hunk 8]
```diff
@@ -29,6 +34,48 @@ def create_from_related_objects(self, *related_objects):
...
+        return model.objects.filter(pk__in=pks)
+
+    def filter_from_related_objects(self, *related_objects):
+        return self.model.objects.filter(
+            Q(template=self.instance) | Q(template__isnull=True)
```
Comment (matllubos) [inline pymess/models/common.py]: self.instance is set only if queryset is used via related object (template). Note for me..

File: pymess/models/common.py (hunk 2 above)
Comment (matllubos) [reply pymess/models/common.py]: please change  the code

File: pymess/models/sms.py (hunk 3 above)
Comment (matllubos) [reply pymess/models/sms.py]: I see it, thanks

File: pymess/migrations/0006_auto_20190322_1712.py (hunk 1 above)
Comment (Formulka) [reply pymess/migrations/0006_auto_20190322_1712.py]: true

File: pymess/models/common.py (hunk 6 above)
Comment (Formulka) [reply pymess/models/common.py]: In that case you assume you are sending all related objects there should be (with mixed content_type), not sure if it makes sense.

File: pymess/models/common.py (hunk 7 above)
Comment (Formulka) [reply pymess/models/common.py]: probably a remnant of something more complex, thank you

File: pymess/models/common.py (hunk 5 above)
Comment (matllubos) [reply pymess/models/common.py]: ?

File: pymess/models/common.py (hunk 6 above)
Comment (matllubos) [reply pymess/models/common.py]: ok let it be

**Pull request merged**